feat(activity): add activity 2

lwasser opened the pull request:
This is the second activity in the workshop. It's complete short of the answers to the OYO sections. i have an answer key script as well that i can push separately.

lwasser commented:
NOTE: the CI issues will be fixed when some of the other pr's are merged. I'm worrying less about that now and more about content!!

ucodery commented on clean-modular-code/activity-2/clean-code-activity-2.md:
```diff
@@ -11,22 +11,37 @@ kernelspec:
+
+## Remember the strategies for cleaner code 
+
+* Document what the code does by adding a docstring at the top of the script to help your future self, a future lab mate, or another user understand that tasks that your code performs.
```
```suggestion
* Document what the code does by adding a docstring at the top of the script to help your future self, a future lab mate, or another user understand the tasks that your code performs.
```

ucodery commented on clean-modular-code/activity-2/clean-code-activity-2.md:
```diff
@@ -36,15 +51,19 @@ What tools could make the code more DRY?
 ## Reproducibility: will this code run on other machines?
 
-One big challenge when considering the reproducibility of workflows is ensuring that your code runs seamlessly on different machines or environments. One key issue is file paths. Hard-coded paths, like "data/part-1-data.json", can lead to errors if the directory structure on another machine differs from your setup. Paths should be constructed dynamically, using tools like Python’s pathlib or envi ...
+One big challenge when considering the reproducibility of workflows is ensuring that your code runs seamlessly on different machines or environments. One key issue is file paths. Hard-coded paths, like `data/part-1-data.json`, can lead to errors if the directory structure on another machine differs from your setup. Paths should be constructed dynamically, using tools like Python’s `pathlib` or en ...
```
I definitely thought you were leading up to Window vs POSIX path slashes. I think that is a more basic, and easier to automate, difference between environments. If the directory structure actually changes in new environments, I think there is only so much you can do without needing user input.

lwasser replied:
Oh - @ucodery i didn't understand this comment at first but now i think i do. 

this section IS about windows vs POSIC path slashes. I think my language that says "directory structure on another machine differs" is confusing / incorrect!!

that is a much harder problem that we can't solve with code as you point out. How about i edit this paragraph to make it more clear. it's fully about path creation which we can handle with code.

ucodery commented on clean-modular-code/activity-2/clean-code-activity-2.md:
```diff
@@ -157,22 +190,36 @@ combined_papers_df.shape
 
-"package-name: title here." 
+1. Add a processing step to the code that pulls the package name out from the title column and creates a new column called `package-name`.
+1. Create a workflow that calculates how many packages in that list are pyOpenSci packages. To calculate how many are pyOpenSci packages,   you will need the `pyos-joss-accepted-packages.csv` file in the `data/` directory with the activity 2 JSON data. This .csv file lists all pyOpenSci packages that have also gone through JOSS review.
```
```suggestion
1. Create a workflow that calculates how many packages in that list are pyOpenSci packages. To calculate how many are pyOpenSci packages, you will need the `pyos-joss-accepted-packages.csv` file in the `data/` directory with the activity 2 JSON data. This .csv file lists all pyOpenSci packages that have also gone through JOSS review.
```

lwasser commented:
Cool - CI is failing with things are are fixed in #18 #13  i am going to merge this soon after one last round of revisions and adding answers to the OYOs

ucodery reviewed (approved):
I just read this one section, I couldn't easily tell if anything else has changed

lwasser commented on clean-modular-code/activity-2/clean-code-activity-2.md:
```diff
@@ -0,0 +1,232 @@
+
+## Reproducibility: will your code run on other machines?
+
+When considering workflow reproducibility, a key challenge is ensuring that your code runs seamlessly across different machines. Hard-coded paths, like data/part-1-data.json, can cause errors if created on a Windows machine but run on a Mac (Unix-based/POSIX) machine.
```
oh you're right! that's an important nit!!

lwasser commented:
thank you, @ucodery. If you want to have a look at a preview of the published (ci) page - it's here - https://output.circle-artifacts.com/output/job/b18a4bf1-8f83-4764-a52b-e0e61a07f909/artifacts/0/html/clean-modular-code/activity-2/clean-code-activity-2.html I've made small edits and clarified things which is why it may be hard to identify what has changed.

The big change here is that I rewrote the POSIX path section and added examples. I'm hopeful that makes it more clear!  Thank you for the review comments - things were not worded well before.

lwasser commented:
I am going to merge this PR, rebase and focus more time on activity 3 next week. Thank you again @ucodery for all of the input!! the CI issues are related to other lessons not in this PR. They will all get fixed once things are merged! Lesson dev is iterative as we focus on the content and move things around accordingly.
